views-jte/src/main/java/io/micronaut/views/jte/PlainJteViewsRenderer.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2017-2021 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.views.jte;
+
+import gg.jte.ContentType;
+import gg.jte.TemplateOutput;
+import gg.jte.html.HtmlTemplateOutput;
+import io.micronaut.core.annotation.NonNull;
+import io.micronaut.http.MediaType;
+import io.micronaut.http.annotation.Produces;
+import io.micronaut.views.ViewsConfiguration;
+import jakarta.inject.Singleton;
+
+import java.io.Writer;
+import java.nio.file.Paths;
+
+/**
+ * JTE renderer constrained to text/plain.
+ *
+ * @param <T> type of input model.
+ * @author edward3h
+ * @since 2021-11-25
+ */
+@Produces(MediaType.TEXT_PLAIN)
+@Singleton
+public class PlainJteViewsRenderer<T> extends JteViewsRenderer<T> {
+    protected PlainJteViewsRenderer(ViewsConfiguration viewsConfiguration, JteViewsRendererConfiguration jteViewsRendererConfiguration) {
+        super(viewsConfiguration, jteViewsRendererConfiguration, ContentType.Plain, Paths.get("build/jte-classes" +
+                                                                                                      "/plain"));
+    }
+
+    @Override
+    @NonNull
+    protected TemplateOutput getOutput(Writer out) {
+        TemplateOutput result = super.getOutput(out);
+        if (!(result instanceof HtmlTemplateOutput)) {
+            result = new PlainHtmlTemplateOutput(result);
+        }
+        return result;
+    }
+
+    /**
+     * workaround to render plain text from a template that was compiled as HTML.
+     */
+    private static class PlainHtmlTemplateOutput implements HtmlTemplateOutput {
```

Comment:
Is this workaround necessary because of a problem with Micronaut or JTE ?

Reply:
JTE has 'Html' and 'Plain' modes which affect how it compiles and renders templates. It's not really designed to do both in one engine so I've set up the renderers to have one of each. In dynamic mode they generate and compile in separate directories.

When the plugin is used to precompile templates, you just get one class for which mode the engine was set to in gradle, which defaults to Html. The workaround here lets the Plain renderer use a class that was compiled in Html mode. (The other way around works already due to class hierachy)

Micronaut Views is selecting templates based on the combination of name and response content-type, it's not all one or the other like JTE.

I'm not sure in practice it makes much sense to render the same template as both Html and Plain. I'll think about whether there's another way to do this.

Reply:
I edited the comment on this class to try and make it explain better.